onnx/onnx-operators.in.proto

```diff
@@ -26,6 +26,55 @@ import "onnx.proto";
 // that describes the ONNX standard operators.
 //
 
+// Operator/function status.
+enum OperatorStatus {
+    EXPERIMENTAL = 0;
+    STABLE = 1;
+}
+
+// Defines an attribute declaration.
+message AttrDeclProto {
+
+  // The attribute name. Must be present.
+  optional string name = 1;
+
+  // The type of this attribute.
+  optional AttributeProto.AttributeType type = 2;
+};
+
+message FunctionProto {
+  // The name of the function, similar usage of op_type in OperatorProto.
+  optional string name = 1;
+  
+  // The first version of a function set which contains this function.
+  // When there's any breaking change for this function, the function set
+  // contains the function needs to bump its version, and since_version of
+  // the updated function will be changed to the updated function set version.  
+  optional int64 since_version = 2;
+
+  // This field indicates whether the syntax, semantics, or presence
+  // of this function is in an experimental or stable stage. Once an
+  // function is published as STABLE, its syntax and semantics MUST NOT
+  // change in subsequent versions of the operator set.
+  // When a function is published as EXPERIMENTAL, the syntax and semantics
+  // of the function MAY change across operator set versions.
+  // Functions "become" stable by deprecating the experimental version and
+  // introducing a new stable function with the same name.
+  optional OperatorStatus status = 3;
+
+  // The inputs and outputs of the function.
+  repeated string input = 4;
+  repeated string output = 5;
+
+  // The attributes declaration of the function.
+  repeated AttrDeclProto attr_decl = 6;
```

Comment:
If we can derive the types of inputs and outputs through inference, why can't we do the same for attributes?

Reply:
Good question. We can. so that the attribute here will also be just "names", which will be referred by node inside the body.